substreams/ethereum-uniswap-v4/src/modules/uni_math.rs

```diff
@@ -0,0 +1,332 @@
+
+    // Apply the mask using XOR and subtraction to restore the sign
+    Ok((amount_0 ^ &mask) - mask)
+}
+
+/// Gets the amount0 delta between two prices
+///
+/// Calculates liquidity / sqrt(lower) - liquidity / sqrt(upper),
+/// i.e. liquidity * (sqrt(upper) - sqrt(lower)) / (sqrt(upper) * sqrt(lower))
+///
+/// ## Arguments
+///
+/// * `sqrt_ratio_a_x96`: A sqrt price assumed to be lower otherwise swapped
+/// * `sqrt_ratio_b_x96`: Another sqrt price
+/// * `liquidity`: The amount of usable liquidity
+/// * `round_up`: Whether to round the amount up or down
+///
+/// ## Returns
+///
+/// Amount of token0 required to cover a position of size liquidity between the two passed prices
+fn get_amount_0_delta(
+    sqrt_ratio_a_x96: BigInt,
+    sqrt_ratio_b_x96: BigInt,
+    liquidity: u128,
+    round_up: bool,
+) -> Result<BigInt, MathError> {
+    let (sqrt_ratio_a_x96, sqrt_ratio_b_x96) = if sqrt_ratio_a_x96 < sqrt_ratio_b_x96 {
+        (sqrt_ratio_a_x96, sqrt_ratio_b_x96)
+    } else {
+        (sqrt_ratio_b_x96, sqrt_ratio_a_x96)
+    };
+
+    if sqrt_ratio_a_x96 == BigInt::from(0) {
+        return Err(MathError::InvalidPrice);
+    }
+
+    let numerator_1 = BigInt::from(liquidity) << 96;
+    let numerator_2 = &sqrt_ratio_b_x96 - &sqrt_ratio_a_x96;
+
+    if round_up {
+        // For rounding up: ceil(ceil(numerator_1 * numerator_2 / sqrt_ratio_b_x96) /
+        // sqrt_ratio_a_x96)
+        let temp =
+            (&numerator_1 * &numerator_2 + &sqrt_ratio_b_x96 - BigInt::from(1)) / &sqrt_ratio_b_x96;
+        Ok((&temp + &sqrt_ratio_a_x96 - BigInt::from(1)) / sqrt_ratio_a_x96)
+    } else {
+        // For rounding down: floor(floor(numerator_1 * numerator_2 / sqrt_ratio_b_x96) /
+        // sqrt_ratio_a_x96)
+        Ok((&numerator_1 * &numerator_2) / &sqrt_ratio_b_x96 / sqrt_ratio_a_x96)
+    }
+}
+
+const Q96: u128 = 1 << 96;
+
+/// Helper that gets signed token1 delta
+///
+/// ## Arguments
+///
+/// * `sqrt_ratio_a_x96`: A sqrt price
+/// * `sqrt_ratio_b_x96`: Another sqrt price
+/// * `liquidity`: The change in liquidity for which to compute the amount1 delta
+///
+/// ## Returns
+///
+/// Amount of token1 corresponding to the passed liquidityDelta between the two prices
+fn get_amount_1_delta_signed(
+    sqrt_ratio_a_x96: BigInt,
+    sqrt_ratio_b_x96: BigInt,
+    liquidity: i128,
+) -> Result<BigInt, MathError> {
+    let sign = !liquidity.is_negative();
+
+    // Create mask for negative numbers
+    let mask = if sign { 0u128 } else { u128::MAX };
+
+    // Get absolute value of liquidity using XOR and addition
+    let liquidity = mask ^ mask.wrapping_add_signed(liquidity);
+
+    // Convert mask to BigInt (all 1s or all 0s)
+    let mask = if sign { BigInt::from(0) } else { -BigInt::from(1) };
+
+    let amount_1 = get_amount_1_delta(sqrt_ratio_a_x96, sqrt_ratio_b_x96, liquidity, sign)?;
+
+    // Apply the mask using XOR and subtraction to restore the sign
+    Ok((amount_1 ^ &mask) - mask)
```

Comment:
I think we can easily remove all bitwise operation there, using just normal functions like `abs()` and `neg()` or 

```
Ok(if sign { amount_1 } else { -amount_1 })
```

Reply:
Ah I wouldn't like to touch it, as it's an accurate representation of the protocol's math